README.md
## Infinite Card Rotation
Flutter widget which gives a flip animation to It's child widgets passed to it as the front and back child.


# Installation

1.  Add the latest version of package to your pubspec.yaml (and run `dart pub get`):
 ```sh
  dependencies:
    infinite_card_rotation: ^1.0.3
```
2.  Import the package and use it in your Flutter App.
 ```sh
import 'package:infinite_card_rotation/infinite_card_rotation.dart';
```
## Preview
> **Example:** Horizontal infinite rotation on the card.
>
[![N|Solid](https://github.com/Rohit-joshi-i/infinite-card-rotation/blob/main/assets/gif/example.gif?raw=true)](https://nodesource.com/products/nsolid)

# Example

```sh
FlipCard(  
  onFlipDone: () {  
    print('Rotation Completed');  
  },  
 front: Container(  
    height: 200,  
  width: 200,  
  decoration: BoxDecoration(  
      shape: BoxShape.circle,  
  color: Colors.red,  
  ),  
  child: Icon(Icons.home,size: 50,),  
  ),  
  back: Container(  
    height: 200,  
  width: 200,  
  child: Icon(Icons.call,size: 50,),  
  decoration: BoxDecoration(  
      color: Colors.blue,  
  shape: BoxShape.circle,),  
  ),  
  flipCount: 10,  
  speed: 500,  
)



CHANGELOG.md
## 1.0.3

* Improved documentation

test/infinite_card_rotation_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:infinite_card_rotation/infinite_card_rotation.dart';

void main() {
  test('adds one to input values', () {
    final calculator = FlipCard(
      front: Container(),
      back: Container(),
    );
  });
}


lib/infinite_card_rotation.dart
library infinite_card_rotation;

import 'dart:async';
import 'dart:math';
import 'package:flutter/material.dart';

enum CardFlipDirection {
  vertical,
  horizontal,
}

class FlipCard extends StatefulWidget {
  final Widget front;
  final Widget back;

  ///
  /// Speed in milliseconds at which the cards with rotate
  ///
  final int speed;

  ///
  /// Direct of the rotation
  ///
  final CardFlipDirection direction;

  ///
  /// Function called when card is Flipped
  ///
  final VoidCallback? onFlip;

  ///
  /// Function called when card flip is completed
  ///
  final VoidCallback? onFlipDone;

  ///
  /// No of times the card is flipped
  ///
  final double flipCount;

  ///
  /// card is flipped infinite times
  ///
  final bool isInfinite;
  final Alignment alignment;

  const FlipCard({
    Key? key,
    required this.front,
    required this.back,
    this.speed = 250,
    this.onFlip,
    this.onFlipDone,
    this.isInfinite = false,
    this.direction = CardFlipDirection.horizontal,
    this.alignment = Alignment.center,
    this.flipCount = 1,
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() => _FlipCardState();
}

class _FlipCardState extends State<FlipCard>
    with SingleTickerProviderStateMixin {
  AnimationController? controller;
  Animation<double>? _firstRotation;
  Animation<double>? _secondRotation;

  late bool isFront;
  late double flipCount;

  @override
  void initState() {
    super.initState();

    ///
    ///
    /// Initialize the late arguments in the init state
    ///
    ///
    isFront = true;
    flipCount = widget.flipCount;
    controller = AnimationController(
      value: isFront ? 0.0 : 1.0,
      duration: Duration(milliseconds: widget.speed),
      vsync: this,
    );

    ///
    ///
    /// First rotation Clockwise
    ///
    ///
    _firstRotation = TweenSequence(
      [
        TweenSequenceItem<double>(
          tween: Tween(begin: 0.0, end: pi / 2)
              .chain(CurveTween(curve: Curves.easeIn)),
          weight: 50.0,
        ),
        TweenSequenceItem<double>(
          tween: ConstantTween<double>(pi / 2),
          weight: 50.0,
        ),
      ],
    ).animate(controller!);

    ///
    ///
    /// Second rotation Clockwise
    ///
    ///
    _secondRotation = TweenSequence(
      [
        TweenSequenceItem<double>(
          tween: ConstantTween<double>(pi / 2),
          weight: 50.0,
        ),
        TweenSequenceItem<double>(
          tween: Tween(begin: pi / 2, end: 0.0)
              .chain(CurveTween(curve: Curves.easeOut)),
          weight: 50.0,
        ),
      ],
    ).animate(controller!);
  }

  /// Flip the card
  Future<void> flipCard() async {
    if (!mounted) return;

    if (widget.isInfinite) {
      widget.onFlip?.call();
      final animation = isFront ? controller!.forward() : controller!.reverse();
      animation.whenComplete(() {
        isFront = !isFront;
        flipCard();
      });
      return;
    }
    widget.onFlip?.call();

    controller!.duration = Duration(milliseconds: widget.speed);

    if (flipCount > 0) {
      final animation = isFront ? controller!.forward() : controller!.reverse();
      flipCount--;
      animation.whenComplete(() {
        isFront = !isFront;
        if (flipCount > 0) {
          flipCard();
        } else {
          flipCount = widget.flipCount;
        }
      });
    }

    ///
    /// If FLIP count reaches 0 call OnFlipDone
    ///
    if (flipCount <= 0) {
      widget.onFlipDone?.call();
      return;
    }
  }

  @override
  Widget build(BuildContext context) {
    final child = Stack(
      alignment: widget.alignment,
      fit: StackFit.passthrough,
      children: <Widget>[
        _buildCard(front: true),
        _buildCard(front: false),
      ],
    );

    return GestureDetector(
      behavior: HitTestBehavior.translucent,
      onTap: widget.flipCount == 0 ? null : flipCard,
      child: child,
    );
  }

  Widget _buildCard({required bool front}) {
    return FlipAnimationCard(
      animation: front ? _firstRotation : _secondRotation,
      direction: widget.direction,
      child: front ? widget.front : widget.back,
    );
  }

  @override
  void dispose() {
    controller!.dispose();
    super.dispose();
  }
}

class FlipAnimationCard extends StatelessWidget {
  const FlipAnimationCard(
      {this.child, this.animation, this.direction, Key? key})
      : super(key: key);

  final Widget? child;
  final Animation<double>? animation;
  final CardFlipDirection? direction;

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: animation!,
      builder: (BuildContext context, Widget? child) {
        var transform = Matrix4.identity();
        if (direction == CardFlipDirection.vertical) {
          transform.rotateX(animation!.value);
        } else {
          transform.rotateY(animation!.value);
        }
        return Transform(
          transform: transform,
          alignment: FractionalOffset.center,
          child: child,
        );
      },
      child: child,
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:infinite_card_rotation/infinite_card_rotation.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            FlipCard(
              onFlipDone: () {
              },
              front: Container(
                height: 200,
                width: 200,
                decoration: const BoxDecoration(
                  shape: BoxShape.circle,
                  color: Colors.red,
                ),
                child: const Icon(
                  Icons.home,
                  size: 50,
                ),
              ),
              back: Container(
                height: 200,
                width: 200,
                decoration: const BoxDecoration(
                  color: Colors.blue,
                  shape: BoxShape.circle,
                ),
                child: const Icon(
                  Icons.call,
                  size: 50,
                ),
              ),
              flipCount: 10,
              speed: 500,
            )
          ],
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
